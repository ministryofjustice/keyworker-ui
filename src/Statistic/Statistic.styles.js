import styled from 'styled-components'
import { SPACING, FONT_SIZE, BREAKPOINTS } from '@govuk-react/constants'

export const Container = styled('div')`
  display: flex;
  flex-direction: column;
  height: 100%;
`

export const Heading = styled('h2')`
  flex: 1 0 auto;
  margin: 0 0 ${SPACING.SCALE_3};
  font-size: ${FONT_SIZE.SIZE_16};

  @media (min-width: ${BREAKPOINTS.LARGESCREEN}) {
    font-size: ${FONT_SIZE.SIZE_19};
  }
`

export const Value = styled('strong')`
  display: block;
  margin: 0 0 ${SPACING.SCALE_3};
  font-size: ${FONT_SIZE.SIZE_24};
  font-weight: bold;
  line-height: 1;

  @media (min-width: ${BREAKPOINTS.LARGESCREEN}) {
    font-size: 36px;
  }
`
