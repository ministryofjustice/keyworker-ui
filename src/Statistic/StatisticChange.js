import React, { Fragment } from 'react'
import PropTypes from 'prop-types'
import { createValueString } from '../stringUtils'

const StatisticChange = ({ change, type }) => {
  const { value, period } = change
  const changeType = value > 0 ? 'increase' : 'decrease'

  return (
    <Fragment>
      {changeType === 'increase' && '+'}
      {createValueString(value, type)}
      <img src={`/images/icon-${changeType}.png`} alt={changeType} height={20} width={20} />
      {period && `since last ${period}`}
    </Fragment>
  )
}

StatisticChange.propTypes = {
  change: PropTypes.shape({
    value: PropTypes.number,
    period: PropTypes.string,
  }).isRequired,
  type: PropTypes.string.isRequired,
}

export default StatisticChange
