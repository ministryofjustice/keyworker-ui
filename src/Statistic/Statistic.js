import React from 'react'
import PropTypes from 'prop-types'
import { Container, Heading, Value, Change } from './Statistic.styles'
import StatisticChange from './StatisticChange'
import { createValueString } from '../stringUtils'

const Statistic = ({ heading, value, change, type }) => (
  <Container data-qa="keyworker-stat">
    <Heading>{heading}</Heading>
    <Value>{createValueString(value, type)}</Value>
    <Change>
      {change.value === 0 && `no change since last ${change.period}`}
      {change.value !== 0 && <StatisticChange change={change} type={type} />}
    </Change>
  </Container>
)

Statistic.propTypes = {
  heading: PropTypes.string,
  value: PropTypes.number,
  change: PropTypes.shape({
    value: PropTypes.number,
    period: PropTypes.string,
  }),
  type: PropTypes.string,
}

Statistic.defaultProps = {
  heading: 'Unknown statistic',
  value: 0,
  change: {
    value: 0,
    period: 'week',
  },
  type: '',
}

export default Statistic
