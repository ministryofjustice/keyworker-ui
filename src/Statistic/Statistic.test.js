import React from 'react'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'
import Statistic from './Statistic'

describe('<Statistic />', () => {
  it('renders without crashing', () => {
    shallow(<Statistic />)
  })

  it('should match the default snapshot', () => {
    const tree = renderer.create(<Statistic />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  describe('with no change in value over the change period', () => {
    const statWithNoChange = {
      heading: 'A statistic with no change over the change period',
      value: 1,
      change: {
        value: 0,
        period: 'week',
      },
      name: 'statWithNoChange',
    }

    it('should match the no change snapshot', () => {
      const tree = renderer.create(<Statistic {...statWithNoChange} />).toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  describe('with positive change value', () => {
    const statWithPositiveChange = {
      heading: 'A statistic with a positive change',
      value: 2,
      change: {
        value: 1,
        period: 'week',
      },
      name: 'statWithPositiveChange',
    }

    it('should match the positive change snapshot', () => {
      const tree = renderer.create(<Statistic {...statWithPositiveChange} />).toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  describe('with negative change value', () => {
    const statWithNegativeChange = {
      heading: 'A statistic with a negative change',
      value: 0,
      change: {
        value: -1,
        period: 'week',
      },
      name: 'statWithNegativeChange',
    }

    it('should match the negative change snapshot', () => {
      const tree = renderer.create(<Statistic {...statWithNegativeChange} />).toJSON()
      expect(tree).toMatchSnapshot()
    })
  })

  describe('with percentages', () => {
    const stateWithPercentageValue = {
      heading: 'A statistic with a percentage',
      value: 5,
      type: 'percentage',
      change: {
        value: 0,
        period: 'week',
      },
      name: 'stateWithPercentageValue',
    }

    it('should decorate the value with a percent', () => {
      const tree = renderer.create(<Statistic {...stateWithPercentageValue} />).toJSON()
      expect(tree).toMatchSnapshot()
    })
  })
})
