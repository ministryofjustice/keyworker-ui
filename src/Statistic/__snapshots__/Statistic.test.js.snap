// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Statistic /> should match the default snapshot 1`] = `
<div
  className="css-flq1iw"
  data-qa="keyworker-stat"
>
  <h2
    className="css-1uz4jaw"
  >
    Unknown statistic
  </h2>
  <strong
    className="css-uyt74n"
  >
    0
  </strong>
  <p
    className="css-1umua3x"
  >
    no change since last week
  </p>
</div>
`;

exports[`<Statistic /> with negative change value should match the negative change snapshot 1`] = `
<div
  className="css-flq1iw"
  data-qa="keyworker-stat"
>
  <h2
    className="css-1uz4jaw"
  >
    A statistic with a negative change
  </h2>
  <strong
    className="css-uyt74n"
  >
    0
  </strong>
  <p
    className="css-1umua3x"
  >
    -1
    <img
      alt="decrease"
      height={20}
      src="/images/icon-decrease.png"
      width={20}
    />
    since last week
  </p>
</div>
`;

exports[`<Statistic /> with no change in value over the change period should match the no change snapshot 1`] = `
<div
  className="css-flq1iw"
  data-qa="keyworker-stat"
>
  <h2
    className="css-1uz4jaw"
  >
    A statistic with no change over the change period
  </h2>
  <strong
    className="css-uyt74n"
  >
    1
  </strong>
  <p
    className="css-1umua3x"
  >
    no change since last week
  </p>
</div>
`;

exports[`<Statistic /> with percentages should decorate the value with a percent 1`] = `
<div
  className="css-flq1iw"
  data-qa="keyworker-stat"
>
  <h2
    className="css-1uz4jaw"
  >
    A statistic with a percentage
  </h2>
  <strong
    className="css-uyt74n"
  >
    5%
  </strong>
  <p
    className="css-1umua3x"
  >
    no change since last week
  </p>
</div>
`;

exports[`<Statistic /> with positive change value should match the positive change snapshot 1`] = `
<div
  className="css-flq1iw"
  data-qa="keyworker-stat"
>
  <h2
    className="css-1uz4jaw"
  >
    A statistic with a positive change
  </h2>
  <strong
    className="css-uyt74n"
  >
    2
  </strong>
  <p
    className="css-1umua3x"
  >
    +
    1
    <img
      alt="increase"
      height={20}
      src="/images/icon-increase.png"
      width={20}
    />
    since last week
  </p>
</div>
`;
